### PR TITLE
Step 5: Admin panel

### PR DESCRIPTION
**Pull request by vtodorchuk**

### CODE REVIEW

File: app/admin/authors.rb [hunk 1]
```diff
@@ -0,0 +1,16 @@
+ActiveAdmin.register Author do
+  permit_params :first_name, :last_name
+
+  filter :books, collection: BookDecorator.decorate_collection(Book.all)
```
Comment (CaptainJNS) [inline app/admin/authors.rb]: why do you need a decorator here?

File: app/admin/books.rb [hunk 2]
```diff
@@ -0,0 +1,61 @@
+ActiveAdmin.register Book do
+  permit_params :title, :description, :price_cents, :dimensions, :year_of_publication, :materials, :quantity,
+                category_ids: [], author_ids: []
+
+  decorate_with BookDecorator
+
+  filter :authors, collection: AuthorDecorator.decorate_collection(Author.all)
```
Comment (CaptainJNS) [inline app/admin/books.rb:7]: and here

File: app/admin/books.rb [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+    column :price_cents, &:price
+    column :quantity
+    column :authors, &:authors_names
+    column :categories, &:categories_names
+
+    actions
+  end
+
+  show do
+    h1 book.title
```
Comment (CaptainJNS) [inline app/admin/books.rb]: doesn't AA works without this line? It is supposed to parse `name/title` columns for title

File: app/admin/books.rb [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+      f.input :title
+      f.input :year_of_publication
+      f.input :price_cents
+      f.input :materials
+      f.input :dimensions
+      f.input :quantity
+      f.input :description, as: :text
+
+      f.input :categories, as: :check_boxes
+      f.input :authors, as: :check_boxes, collection: AuthorDecorator.decorate_collection(Author.all)
```
Comment (CaptainJNS) [inline app/admin/books.rb]: Why checkboxes but not a `select`?
And what about collection for `categories`?

File: app/admin/categories.rb [hunk 5]
```diff
@@ -0,0 +1,6 @@
+ActiveAdmin.register Category do
+  permit_params :name, book_ids: []
+
+  filter :books, collection: BookDecorator.decorate_collection(Book.all)
```
Comment (CaptainJNS) [inline app/admin/categories.rb]: Why decorator?

File: app/admin/dashboard.rb [hunk 6]
```diff
@@ -0,0 +1,34 @@
+# frozen_string_literal: true
+
+ActiveAdmin.register_page 'Dashboard' do
```
Comment (CaptainJNS) [inline app/admin/dashboard.rb:3]: remove this page

File: config/initializers/will_paginate.rb [hunk 7]
```diff
@@ -0,0 +1,21 @@
+if defined?(WillPaginate)
```
Comment (CaptainJNS) [inline config/initializers/will_paginate.rb]: This is not an initializer, this is a monkey patch. Why do you need it?

File: spec/factories/admin_users.rb [hunk 8]
```diff
@@ -0,0 +1,6 @@
+FactoryBot.define do
+  factory :admin_user do
+    email { "admin-#{FFaker::Name.last_name}@admin.ua" }
```
Comment (CaptainJNS) [inline spec/factories/admin_users.rb]:
```suggestion
    email { FFaker::Internet.email }
```

File: spec/feature/admin/admin_author_spec.rb [hunk 9]
```diff
@@ -0,0 +1,63 @@
+require 'rails_helper'
+require 'support/shared_context'
+
+RSpec.describe 'Admin_Author' do
```
Comment (CaptainJNS) [inline spec/feature/admin/admin_author_spec.rb]:
```suggestion
RSpec.describe 'Author' do
```
And rename file `/authors.spec`. The same for other admin pages tests

File: spec/feature/admin/admin_author_spec.rb [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+    visit 'admin/authors'
+
+    click_link('New Author')
+
+    fill_in 'author[first_name]', with: first_name
+    fill_in 'author[last_name]', with: last_name
+
+    click_button('Create Author')
+
+    expect(page).to have_content(first_name)
```
Comment (CaptainJNS) [inline spec/feature/admin/admin_author_spec.rb]:
```suggestion
    expect { click_button('Create Author') }.to change(Author, :count).from(0).to(1)
```

File: spec/feature/admin/admin_author_spec.rb [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+  end
+
+  it 'can delete an author' do
+    author
+
+    visit 'admin/authors'
+
+    click_link('Delete')
+
+    expect(page).not_to have_content(author.first_name)
```
Comment (CaptainJNS) [inline spec/feature/admin/admin_author_spec.rb]:
```suggestion
    expect { click_link('Delete') }.to change(Author, :count).from(1).to(0)
```

File: spec/feature/admin/admin_author_spec.rb [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+
+    visit 'admin/authors'
+
+    click_link('Edit')
+
+    fill_in 'author[first_name]', with: 'Vlad'
+
+    click_button('Update Author')
+
+    expect(page).to have_content('Vlad')
```
Comment (CaptainJNS) [inline spec/feature/admin/admin_author_spec.rb]:
```suggestion
    expect { click_button('Update Author') }.to change { author.reload.first_name }.from(old_name).to(new_name)
```

File: spec/models/user_spec.rb [hunk 13]
```diff
@@ -1,6 +1,8 @@
 require 'rails_helper'
 
 RSpec.describe User, type: :model do
+  let(:user) { FactoryBot.create(:user) }
```
Comment (CaptainJNS) [inline spec/models/user_spec.rb]: do you need it?

File: spec/support/shared_context.rb [hunk 14]
```diff
@@ -9,8 +9,8 @@
 end
 
 RSpec.shared_context 'with api request authentication helper methods' do
-  def sign_in(user)
-    login_as(user, scope: :user)
+  def sign_in(user, user_type = :user)
+    login_as(user, scope: user_type)
```
Comment (CaptainJNS) [inline spec/support/shared_context.rb]:
```suggestion
  def sign_in(user, scope = :user)
    login_as(user, scope: scope)
```

**Review by CaptainJNS — CHANGES_REQUESTED** [review]

File: app/admin/books.rb (hunk 2 above)
Comment (vtodorchuk) [reply app/admin/books.rb:7]: In order to print author full_name i used decorator.

File: app/admin/dashboard.rb (hunk 6 above)
Comment (vtodorchuk) [reply app/admin/dashboard.rb:3]: In next steps i use this page for SideKiq button that brings admin to /sidekiq page. I dont know if i doing right by it was not really comfortable when i was working with sidekiq to check if my task are running well so i use dashboard for this purposes.

File: app/admin/authors.rb [hunk 15]
```diff
@@ -0,0 +1,35 @@
...
+      row :books
+    end
+  end
+
+  form do |f|
+    f.inputs do
+      f.input :first_name
+      f.input :last_name
+
+      f.input :books, as: :select, collection: Book.all
```
Comment (CaptainJNS) [inline app/admin/authors.rb]: I don't think we need to attach books to authors. It has to be vice versa

File: app/admin/categories.rb [hunk 16]
```diff
@@ -0,0 +1,31 @@
...
+      row :name
+      row :created_at
+      row :updated_at
+    end
+  end
+
+  form do |f|
+    f.inputs do
+      f.input :name
+      f.input :books, as: :select, collection: Book.all
```
Comment (CaptainJNS) [inline app/admin/categories.rb]: The same as with authors. Do not attach books here

File: app/decorators/comment_decorator.rb [hunk 17]
```diff
@@ -0,0 +1,4 @@
+class CommentDecorator < Draper::Decorator
+  delegate_all
+  decorates_association :book
+end
```
Comment (CaptainJNS) [inline app/decorators/comment_decorator.rb]: Do you decorate comments?

File: spec/feature/admin/book_spec.rb [hunk 18]
```diff
@@ -0,0 +1,61 @@
...
+    fill_in 'book[description]', with: params[:description]
+    fill_in 'book[year_of_publication]', with: params[:year_of_publication]
+    fill_in 'book[price_cents]', with: params[:price_cents]
+    fill_in 'book[materials]', with: params[:materials]
+    fill_in 'book[dimensions]', with: params[:dimensions]
+    fill_in 'book[quantity]', with: params[:quantity]
+
+    click_button('Create Book')
+
+    expect(page).to have_content(params[:title])
```
Comment (CaptainJNS) [inline spec/feature/admin/book_spec.rb]: Use `expect{}.to change{}`. The same as in `author_spec` tests. Check everywhere

File: spec/feature/new_password_spec.rb [hunk 19]
```diff
@@ -0,0 +1,24 @@
...
+
+RSpec.describe 'New Password', type: :feature do
+  let(:user) { FactoryBot.create(:user) }
+
+  describe '#new_password' do
+    it 'change password' do
+      user.reset_password('newPassword', 'newPassword')
+
+      expect(user.password).to eq('newPassword')
+    end
```
Comment (CaptainJNS) [inline spec/feature/new_password_spec.rb]: This is not a feature test

File: spec/feature/user_spec.rb [hunk 20]
```diff
@@ -1,8 +1,6 @@
-require 'rails_helper'
-
 RSpec.describe 'User', type: :feature do
-  let(:user) { FactoryBot.build(:user, email: 'test.rspec@rspec.ua') }
-  let(:user_second) { FactoryBot.create(:user, email: 'test.rspec-test@rspec.ua') }
+  let(:user) { build(:user, email: 'test.rspec@rspec.ua') }
+  let(:user_second) { create(:user, email: 'test.rspec-test@rspec.ua') }
```
Comment (CaptainJNS) [inline spec/feature/user_spec.rb]: why do you need exactly these emails? Why don't you use a FFaker's generations?

**Review by CaptainJNS — CHANGES_REQUESTED** [review]

**Review by CaptainJNS — APPROVED** [review]

**Pull request merged**